Adds linting rule to avoid assignment to 'module' variable

mward-sudo opened the pull request:
Fixes #34859

## Bug

- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [x] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [x] Make sure the linting passes by running `yarn lint`

mward-sudo commented:
Hi, first contribution. Happy to make any suggested changes.

balazsorban44 commented on packages/eslint-plugin-next/lib/rules/no-assign-module-variable.js:
```diff
@@ -0,0 +1,30 @@
+    return {
+      VariableDeclaration(node) {
+        // Checks node.declarations array for variable with id.name of 'module'
+        const moduleVariableFound = node.declarations.find(
```
Nitpick, but semantically, using `.some` [would be better](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find#:~:text=If%20you%20need%20to%20find%20if%20any%20element%20satisfies%20the%20provided%20testing%20function%2C%20use%20Array.prototype.some().) than a `.find` here, which returns a boolean.

mward-sudo replied:
Agreed.

ijjk commented:
## Failing test suites <!--5669137598-->
Commit: b2ddbc100ea80b914f089698de1e905670875d72

`yarn testheadless test/integration/eslint/test/index.test.js` 
- ESLint > Next Build > empty directories do not fail the build
- ESLint > Next Build > eslint ignored directories do not fail the build
- ESLint > Next Build > eslint caching is enabled
- ESLint > Next Build > eslint cache lives in the user defined build directory
- ESLint > Next Lint > success message when no warnings or errors
- ESLint > Next Lint > max warnings flag does not error when warnings do not exceed threshold
- ESLint > Next Lint > format flag supports additional user-defined formats
- ESLint > Next Lint > eslint caching is enabled by default
- ESLint > Next Lint > eslint caching is disabled with the --no-cache flag
- ESLint > Next Lint > the default eslint cache lives in the user defined build directory
- ESLint > Next Lint > the --cache-location flag allows the user to define a separate cache location
- ESLint > Next Lint > the default eslint caching strategy is metadata
- ESLint > Next Lint > cache with content strategy is different from the one with default strategy

<details>
<summary>Expand output</summary>

  ● ESLint › Next Build › empty directories do not fail the build

    expect(received).toContain(expected) // indexOf

    Expected substring: "Compiled successfully"
    Received string:    "info  - Checking validity of types...·
    Failed to compile.·
    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable
    4:5  Warning: External synchronous scripts are forbidden. See: https://nextjs.org/docs/messages/no-sync-scripts  @next/next/no-sync-scripts·
    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
    "

      124 |         'Warning: External synchronous scripts are forbidden'
      125 |       )
    > 126 |       expect(output).toContain('Compiled successfully')
          |                      ^
      127 |     })
      128 |
      129 |     test('eslint ignored directories do not fail the build', async () => {

      at Object.<anonymous> (integration/eslint/test/index.test.js:126:22)

  ● ESLint › Next Build › eslint ignored directories do not fail the build

    expect(received).toContain(expected) // indexOf

    Expected substring: "Compiled successfully"
    Received string:    "info  - Checking validity of types...·
    Failed to compile.·
    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable
    4:5  Warning: External synchronous scripts are forbidden. See: https://nextjs.org/docs/messages/no-sync-scripts  @next/next/no-sync-scripts·
    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
    "

      139 |         'Warning: External synchronous scripts are forbidden'
      140 |       )
    > 141 |       expect(output).toContain('Compiled successfully')
          |                      ^
      142 |     })
      143 |
      144 |     test('missing Next.js plugin', async () => {

      at Object.<anonymous> (integration/eslint/test/index.test.js:141:22)

  ● ESLint › Next Build › eslint caching is enabled

    command failed with code 1
    warn  - No build cache found. Please configure build caching for faster rebuilds. Read more: https://nextjs.org/docs/messages/no-cache
    info  - Checking validity of types...

    Failed to compile.

    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable

    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

      199 |       ) {
      200 |         return reject(
    > 201 |           new Error(`command failed with code ${code}\n${mergedStdio}`)
          |           ^
      202 |         )
      203 |       }
      204 |

      at ChildProcess.<anonymous> (lib/next-test-utils.js:201:11)

  ● ESLint › Next Build › eslint cache lives in the user defined build directory

    command failed with code 1
    warn  - No build cache found. Please configure build caching for faster rebuilds. Read more: https://nextjs.org/docs/messages/no-cache
    info  - Checking validity of types...

    Failed to compile.

    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable

    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

      199 |       ) {
      200 |         return reject(
    > 201 |           new Error(`command failed with code ${code}\n${mergedStdio}`)
          |           ^
      202 |         )
      203 |       }
      204 |

      at ChildProcess.<anonymous> (lib/next-test-utils.js:201:11)

  ● ESLint › Next Lint › success message when no warnings or errors

    expect(received).toContain(expected) // indexOf

    Expected substring: "No ESLint warnings or errors"
    Received string:    "
    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable·
    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
    "

      341 |
      342 |       const output = stdout + stderr
    > 343 |       expect(output).toContain('No ESLint warnings or errors')
          |                      ^
      344 |     })
      345 |
      346 |     test("don't create .eslintrc file if package.json has eslintConfig field", async () => {

      at Object.<anonymous> (integration/eslint/test/index.test.js:343:22)

  ● ESLint › Next Lint › max warnings flag does not error when warnings do not exceed threshold

    expect(received).toEqual(expected) // deep equality

    Expected: ""
    Received: "
    ./pages/about.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable
    4:5  Warning: External synchronous scripts are forbidden. See: https://nextjs.org/docs/messages/no-sync-scripts  @next/next/no-sync-scripts·
    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable
    4:5  Warning: External synchronous scripts are forbidden. See: https://nextjs.org/docs/messages/no-sync-scripts  @next/next/no-sync-scripts·
    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
    "

      441 |       )
      442 |
    > 443 |       expect(stderr).toEqual('')
          |                      ^
      444 |       expect(stderr).not.toContain(
      445 |         'Warning: External synchronous scripts are forbidden'
      446 |       )

      at Object.<anonymous> (integration/eslint/test/index.test.js:443:22)

  ● ESLint › Next Lint › format flag supports additional user-defined formats

    expect(received).toContain(expected) // indexOf

    Expected substring: "<script src=\"https://example.com\" />"
    Received string:    ""

      464 |         'warning: External synchronous scripts are forbidden'
      465 |       )
    > 466 |       expect(stdout).toContain('<script src="https://example.com" />')
          |                      ^
      467 |       expect(stdout).toContain('2 warnings found')
      468 |     })
      469 |

      at Object.<anonymous> (integration/eslint/test/index.test.js:466:22)

  ● ESLint › Next Lint › eslint caching is enabled by default

    command failed with code 1

    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable

    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

      199 |       ) {
      200 |         return reject(
    > 201 |           new Error(`command failed with code ${code}\n${mergedStdio}`)
          |           ^
      202 |         )
      203 |       }
      204 |

      at ChildProcess.<anonymous> (lib/next-test-utils.js:201:11)

  ● ESLint › Next Lint › eslint caching is disabled with the --no-cache flag

    command failed with code 1

    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable

    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

      199 |       ) {
      200 |         return reject(
    > 201 |           new Error(`command failed with code ${code}\n${mergedStdio}`)
          |           ^
      202 |         )
      203 |       }
      204 |

      at ChildProcess.<anonymous> (lib/next-test-utils.js:201:11)

  ● ESLint › Next Lint › the default eslint cache lives in the user defined build directory

    command failed with code 1

    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable

    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

      199 |       ) {
      200 |         return reject(
    > 201 |           new Error(`command failed with code ${code}\n${mergedStdio}`)
          |           ^
      202 |         )
      203 |       }
      204 |

      at ChildProcess.<anonymous> (lib/next-test-utils.js:201:11)

  ● ESLint › Next Lint › the --cache-location flag allows the user to define a separate cache location

    command failed with code 1

    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable

    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

      199 |       ) {
      200 |         return reject(
    > 201 |           new Error(`command failed with code ${code}\n${mergedStdio}`)
          |           ^
      202 |         )
      203 |       }
      204 |

      at ChildProcess.<anonymous> (lib/next-test-utils.js:201:11)

  ● ESLint › Next Lint › the default eslint caching strategy is metadata

    command failed with code 1

    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable

    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

      199 |       ) {
      200 |         return reject(
    > 201 |           new Error(`command failed with code ${code}\n${mergedStdio}`)
          |           ^
      202 |         )
      203 |       }
      204 |

      at ChildProcess.<anonymous> (lib/next-test-utils.js:201:11)

  ● ESLint › Next Lint › cache with content strategy is different from the one with default strategy

    command failed with code 1

    ./pages/index.js
    1:1  Error: Definition for rule '@next/next/no-assign-module-variable' was not found.  @next/next/no-assign-module-variable

    info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules

      199 |       ) {
      200 |         return reject(
    > 201 |           new Error(`command failed with code ${code}\n${mergedStdio}`)
          |           ^
      202 |         )
      203 |       }
      204 |

      at ChildProcess.<anonymous> (lib/next-test-utils.js:201:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

balazsorban44 commented on packages/eslint-plugin-next/lib/index.js:
```diff
@@ -45,6 +45,7 @@ module.exports = {
         '@next/next/no-duplicate-head': 2,
         '@next/next/inline-script-id': 2,
+        '@next/next/no-assign-module-variable': 2,
       },
```
Higher up in this file, you'll need to add to `rules`:

`'no-assign-module-variable': require('./rules/no-assign-module-variable'),`

ijjk commented:
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| buildDuration | 15.7s | 15.5s | -133ms |
| buildDurationCached | 6.1s | 6.2s | ⚠️ +157ms |
| nodeModulesSize | 456 MB | 456 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.12 | 3.128 | ⚠️ +0.01 |
| / avg req/sec | 801.37 | 799.2 | ⚠️ -2.17 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.355 | 1.36 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1845.58 | 1838.29 | ⚠️ -7.29 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.48 kB | 5.48 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.2 kB | 15.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| buildDuration | 19.2s | 19s | -172ms |
| buildDurationCached | 6.1s | 6.2s | ⚠️ +113ms |
| nodeModulesSize | 456 MB | 456 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.109 | 3.117 | ⚠️ +0.01 |
| / avg req/sec | 804.02 | 802.09 | ⚠️ -1.93 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.371 | 1.365 | -0.01 |
| /error-in-render avg req/sec | 1823.95 | 1832.02 | +8.07 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.59 kB | 5.59 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.33 kB | 2.33 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.3 kB | 15.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: b2ddbc100ea80b914f089698de1e905670875d72

ijjk commented:
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| buildDuration | 15.7s | 15.6s | -83ms |
| buildDurationCached | 6.2s | 6.2s | -84ms |
| nodeModulesSize | 456 MB | 456 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.989 | 2.994 | ⚠️ +0.01 |
| / avg req/sec | 836.49 | 835.02 | ⚠️ -1.47 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.215 | 1.232 | ⚠️ +0.02 |
| /error-in-render avg req/sec | 2056.87 | 2028.87 | ⚠️ -28 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.48 kB | 5.48 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.2 kB | 15.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| buildDuration | 19.5s | 19.6s | ⚠️ +64ms |
| buildDurationCached | 6.2s | 6.3s | ⚠️ +122ms |
| nodeModulesSize | 456 MB | 456 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.023 | 3.016 | -0.01 |
| / avg req/sec | 826.95 | 828.93 | +1.98 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.253 | 1.235 | -0.02 |
| /error-in-render avg req/sec | 1995.66 | 2024.95 | +29.29 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.59 kB | 5.59 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.33 kB | 2.33 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.3 kB | 15.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 03ba0c16a50377a9851fb8c937422b21dd2fe293

mward-sudo commented:
Rebased and squashed commits.

I'm seeing test failures that I don't think are related to my changes, as they are occurring on canary branch as well e.g.:

```
> yarn testheadless test/integration/telemetry/test/index.test.js

yarn run v1.22.17
$ cross-env HEADLESS=true yarn testonly test/integration/telemetry/test/index.test.js
$ yarn jest --runInBand test/integration/telemetry/test/index.test.js
$ /Users/michael/Dev/next.js/node_modules/.bin/jest --runInBand test/integration/telemetry/test/index.test.js
 FAIL  test/integration/telemetry/test/index.test.js
  ● Test suite failed to run

    Failed to deserialize argument at `2` as next_swc::TransformOptions
    JSON: {"jsc":{"parser":{"syntax":"ecmascript","dynamicImport":true,"decorators":false,"jsx":true,"importAssertions":true},"experimental":{"keepImportAssertions":true},"transform":{"hidden":{"jest":true},"legacyDecorator":false,"decoratorMetadata":false,"react":{"importSource":"react","runtime":"automatic","pragma":"React.createElement","pragmaFrag":"React.Fragment","throwIfNamespace":true,"development":false,"useBuiltins":true,"refresh":false},"optimizer":{"simplify":false,"globals":null},"regenerator":{"importPath":"/Users/michael/Dev/next.js/packages/next/dist/compiled/regenerator-runtime/runtime.js"}}},"sourceMaps":"inline","styledComponents":null,"emotion":null,"env":{"targets":{"node":"16.13.0"},"include":["proposal-optional-chaining","proposal-nullish-coalescing-operator"]},"module":{"type":"commonjs"},"disableNextSsg":true,"disablePageConfig":true,"filename":"/Users/michael/Dev/next.js/test/integration/telemetry/test/index.test.js"}

    Caused by:
        unknown field `emotion` at line 1 column 947

      166 |         }
      167 |
    > 168 |         return bindings.transformSync(
          |                         ^
      169 |           isModule ? JSON.stringify(src) : src,
      170 |           isModule,
      171 |           toBuffer(options)

      at Object.transformSync (../packages/next/build/swc/index.js:168:25)
      at Object.transformSync (../packages/next/build/swc/index.js:213:19)
      at Object.process (../packages/next/build/swc/jest-transformer.js:55:27)
      at ScriptTransformer.transformSource (../node_modules/@jest/transform/build/ScriptTransformer.js:612:31)
      at ScriptTransformer._transformAndBuildScript (../node_modules/@jest/transform/build/ScriptTransformer.js:758:40)
      at ScriptTransformer.transform (../node_modules/@jest/transform/build/ScriptTransformer.js:815:19)

info  - Using locally built binary of @next/swc
Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.655 s
Ran all test suites matching /test\/integration\/telemetry\/test\/index.test.js/i.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

mward-sudo commented:
> ## Failing test suites
> Commit: [b2ddbc1](https://github.com/vercel/next.js/commit/b2ddbc100ea80b914f089698de1e905670875d72)
> 
> `yarn testheadless test/integration/eslint/test/index.test.js`
> 
> * ESLint > Next Build > empty directories do not fail the build
> * ESLint > Next Build > eslint ignored directories do not fail the build
> * ESLint > Next Build > eslint caching is enabled
> * ESLint > Next Build > eslint cache lives in the user defined build directory
> * ESLint > Next Lint > success message when no warnings or errors
> * ESLint > Next Lint > max warnings flag does not error when warnings do not exceed threshold
> * ESLint > Next Lint > format flag supports additional user-defined formats
> * ESLint > Next Lint > eslint caching is enabled by default
> * ESLint > Next Lint > eslint caching is disabled with the --no-cache flag
> * ESLint > Next Lint > the default eslint cache lives in the user defined build directory
> * ESLint > Next Lint > the --cache-location flag allows the user to define a separate cache location
> * ESLint > Next Lint > the default eslint caching strategy is metadata
> * ESLint > Next Lint > cache with content strategy is different from the one with default strategy
> 
> Expand output
> Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).
> 
> `yarn testheadless test/integration/telemetry/test/index.test.js`
> 
> * Telemetry CLI > emits telemetry for lint during build
> * Telemetry CLI > emits telemetry for usage of swc
> 
> Expand output
> Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

@ijjk these tests are all passing now.

ijjk commented:
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| buildDuration | 15.5s | 15.2s | -361ms |
| buildDurationCached | 6.1s | 6s | -65ms |
| nodeModulesSize | 478 MB | 478 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.001 | 3.034 | ⚠️ +0.03 |
| / avg req/sec | 833.17 | 823.92 | ⚠️ -9.25 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.176 | 1.173 | 0 |
| /error-in-render avg req/sec | 2125.05 | 2131.68 | +6.63 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.3 kB | 28.3 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72 kB | 72 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.05 kB | 3.05 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.73 kB | 5.73 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| buildDuration | 18.3s | 18.1s | -220ms |
| buildDurationCached | 6s | 6s | -63ms |
| nodeModulesSize | 478 MB | 478 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.045 | 3.039 | -0.01 |
| / avg req/sec | 821 | 822.74 | +1.74 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.173 | 1.177 | 0 |
| /error-in-render avg req/sec | 2131.93 | 2123.33 | ⚠️ -8.6 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.7 kB | 28.7 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.6 kB | 72.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.03 kB | 3.03 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.77 kB | 5.77 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.44 kB | 2.44 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | scripthungry/next.js lint-module-variable | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: f4ff4560c77293047de35de652740305d639c6f3

timneutkens commented:
Thanks @mward-sudo 🎉
